radix-engine/src/transaction/transaction_executor.rs

```diff
@@ -243,97 +508,316 @@ where
+            let mut substate: LiquidFungibleResource =
+                track.read_substate(handle).0.as_typed().unwrap();
+            substate.put(
+                collected_fees
+                    .take_by_amount(proposer_rewards + validator_set_rewards)
+                    .unwrap(),
+            );
+            track.update_substate(handle, IndexedScryptoValue::from_typed(&substate));
+            track.release_lock(handle);
+        }
+
+        (fee_summary, fee_payments)
+    }
+
+    fn update_intent_hash_store(
+        track: &mut Track<S, SpreadPrefixKeyMapper>,
+        next_epoch: Epoch,
+        intent_hash: Hash,
+        expiry_epoch: Option<Epoch>,
+        is_success: bool,
+    ) {
+        // Read the intent hash store
+        let handle = track
+            .acquire_lock(
+                INTENT_HASH_STORE.as_node_id(),
+                MAIN_BASE_PARTITION,
+                &IntentHashStoreField::IntentHashStore.into(),
+                LockFlags::MUTABLE,
+            )
+            .unwrap()
+            .0;
+        let mut substate: IntentHashStoreSubstate =
+            track.read_substate(handle).0.as_typed().unwrap();
+
+        // Update the status of the intent hash
+        if let Some(expiry_epoch) = expiry_epoch {
+            if let Some(partition_number) = substate.partition_of(expiry_epoch.number()) {
+                let handle = track
+                    .acquire_lock_virtualize(
+                        INTENT_HASH_STORE.as_node_id(),
+                        PartitionNumber(partition_number),
+                        &SubstateKey::Map(intent_hash.to_vec()),
+                        LockFlags::MUTABLE,
+                        || {
+                            Some(IndexedScryptoValue::from_typed(&KeyValueEntrySubstate {
+                                value: Option::<IntentHashStatus>::None,
+                                mutability: SubstateMutability::Mutable,
+                            }))
+                        },
+                    )
+                    .unwrap()
+                    .0;
+                track.update_substate(
+                    handle,
+                    IndexedScryptoValue::from_typed(&KeyValueEntrySubstate {
+                        value: Some(if is_success {
+                            IntentHashStatus::CommittedSuccess
+                        } else {
+                            IntentHashStatus::CommittedFailure
+                        }),
+                        // TODO: maybe make it immutable, but how does this affect partition deletion?
+                        mutability: SubstateMutability::Mutable,
+                    }),
+                );
+                track.release_lock(handle);
+            }
+        }
+
+        loop {
+            // TODO: how do we align this with TransactionValidator?
+            if substate
+                .partition_of(next_epoch.after(DEFAULT_MAX_EPOCH_RANGE).number())
+                .is_none()
```

Comment:
Ooo I think you're trying to keep a hold of the state for as long as possible?
Hmm I'd personally rather we keep things as simple as possible, and just delete the state as soon as it expires.

Reply:
> I think you're trying to keep a hold of the state for as long as possible?

Partially correct. The main invariant is that the tracked intent hash range is greater than the max header epoch range.

>  just delete the state as soon as it expires.

This isn't as simple as it sounds. It would require scanning of a partition and delete individuals manually.

Reply:
The transaction tracker stores hashes by their expiry epoch exclusive, right? 

So if, when we advance an epoch to epoch N, and this causes the transaction tracker to advance the minimum partition it tracks from P to P + 1, then P must have handled expiry epochs M to N inclusive (for some M < N), and therefore must now be fully expired (ie all the epochs that map into it are in the past) - so that partition can be deleted.

Reply:
I'd personally rather the deletion logic cleans up sensibly/deterministically so we store minimal additional state, and can explain it more easily.